gputop/remote/gputop-nodejs-csv.js

```diff
@@ -20,21 +24,124 @@ GPUTopNodeJSUI.prototype.weblog = function(message)
     console.log(message);
 }
 
-GPUTopNodeJSUI.prototype.display_features = function(message)
+GPUTopNodeJSUI.prototype.update_features = function(features)
 {
+    console.log(features);
+
+    features.supported_oa_query_guids.forEach(function (guid, i, a) {
+        var metric = gputop.get_map_metric(guid);
+        metric.counters_.forEach(function (i, j, k) {
+            metric.counters_[j].record_data = true;
+        });
+    });
+
+    gputop.open_oa_metric_set({guid: features.supported_oa_query_guids[0]});
+    supported_oa_query_guids = features.supported_oa_query_guids;
+}
+
+GPUTopNodeJSUI.prototype.queue_redraw = function()
+{
+    if (supported_oa_query_guids !== undefined)
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+}
+
+GPUTopNodeJSUI.prototype.load_metrics_panel = function(open_query)
+{
+    open_query();
+}
+
+GPUTopNodeJSUI.prototype.update_slider_period = function(period)
+{
+}
+
+GPUTopNodeJSUI.prototype.render_bars = function()
+{
+}
+
+GPUTopNodeJSUI.prototype.log = function(level, message)
+{
+    console.log(level);
     console.log(message);
 }
 
 var gputop_ui = new GPUTopNodeJSUI();
 
-function gputop_ready(gputop)
+function gputop_ready(gputop_inst)
 {
+    gputop = gputop_inst;
     gputop.connect();
 }
 
-http.get("http://localhost:7890/gputop.js", function(response) {
+function gputop_is_demo () {
+    return false;
+}
+
+function on_gputop_web_ready(gputop_web) {
+    http.get("http://localhost:7890/gputop.js", function(response) {
+        var gputopjs = "";
+        response.setEncoding('utf8');
+        response.on('data', function(data) {
+            gputopjs += data;
+        });
+        response.on('end', function () {
+            eval(gputop_web);
+            eval(gputopjs);
+        });
+    });
+}
+
+http.get("http://localhost:7890/gputop-web.js", function(response) {
+    var gputop_web = "";
     response.setEncoding('utf8');
     response.on('data', function(data) {
-        eval(data);
+        gputop_web += data;
+    });
+    response.on('end', function () {
+        on_gputop_web_ready(gputop_web);
     });
 });
+
+function gputop_clean_exit()
+{
+    if (gputop !== undefined && supported_oa_query_guids !== undefined) {
+        var metric = gputop.get_map_metric(supported_oa_query_guids[0]);
+        var fs = require('fs');
+        var stream = fs.createWriteStream("my_file.csv");
+        stream.once('open', function(fd) {
+            var csv_file = "Metric,Counter,Start,End,Value,Maximum,Reason\n";
```

Comment:
the 'Reason' here is an internal book keeping reason, so think we should skip that.

I think here we want to aim to start with give each metric a single value column, maybe expanding to multiple columns per metric for things like start, end, max.

So instead of being like a log of individual updates, I think we want to aim for it to be possible to e.g. open the CSV file in something like LibreOffice and be able to create graphs from any of the columns of metrics.

One of the fiddly parts here is that updates from the webc code come one value at a time, for a single metric. The boundary between rows in the CSV output can be based on changes to the 'Start' value.

In addition to a column per counter value there would also be a 'Timestamp' column which would be the mid-point between Start and End, corresonding to all the per counter values on the same row.

I think it would be good to aim to do this logging as part of a UI.update_counter function as updates are recieved from the webc code so that the CSV data can be streamed to disk as it's recieved, only requiring a flush on error/exit, and only requiring us to buffer a single row of values at a time instead of the full file. Potentially no buffering is required but then maybe there'd be a slightly higher chance of writting an incomplete line at the end of the file on error/exit.